### PR TITLE
AE-990: Governance template

### PR DESCRIPTION
**Pull request by echoAE**
This agent template provides all of the logic necessary for creating alerts for OpenZeppelin Governance contracts (base contracts and extension contracts).  All of the ABI files are present in the repository.

To test, use the following `agent-config.json` file:
```
{
  "developerAbbreviation": "AE",
  "protocolName": "Compound",
  "protocolAbbreviation": "COMP",
  "governance": {
    "address": "0xc0Da02939E1441F497fd74F78cE7Decb17B66529",
    "abiFile": "GovernorTimelockCompound.json"
  }
}
```

The following transactions should generate alerts:
```
// VoteCast Tx: 0x4f8f3d7827467229caffb8570a9b38aa78448ca358863d7ad17ea4a75673745d
// ProposalCreated Tx: 0x46fb3ad1837ca8fbbd88e6654424752faa20430fab0517c2d965b26759292d4b
// ProposalExecuted Tx: 0xd1924e6d2b04303262fe90d6a354f92d795a6ce4bd7908cd77869aae50e77f43
// ProposalQueued Tx: 0x1a9692a039baebcedc28b13447e3528a6a469b8ae9e8241df14f50a5cbad25dd
// ProposalCanceled Tx: 0xc09ea70ee0f2e05eff9ff63b09195a6714681c18a145c019de8e5fbc9e01e675
```

### CODE REVIEW

File: governance/SETUP.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+address under the `address` key and the appropriate name of the ABI file under the `abiFile` key.  All
+of the supported ABI files are already present in the `./abi` subdirectory, so the only step necessary
+is to type the correct name of the applicable ABI file.
+
+6. Create a new README.md file to provide a description of your agent, using examples from the Forta Github
+repository.  Also update the name and description fields in the `package.json` file.
+
+7. Move files to have the following directory structure:
+```
+  monitor-function-calls/
```
Comment (arbitrary-sparrow) [inline governance/SETUP.md:36]:
```suggestion
  governance/
```

File: governance/SETUP.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
+# Governance Events Agent Template
+
+This agent monitors governance contracts that use the modular system of Governance contracts
+available from OpenZepplin.  All of the possible emitted events are coded into the logic of the
```
Comment (arbitrary-sparrow) [inline governance/SETUP.md:4]:
```suggestion
available from OpenZeppelin.  All of the possible emitted events are coded into the logic of the
```

File: governance/package.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "forta-{developerAbbreviation}-{protocolAbbreviation}-function-calls",
+  "version": "0.0.1",
+  "description": "Forta Agent template for monitoring function calls from {protocolName} contracts",
```
Comment (arbitrary-sparrow) [inline governance/package.json:4]: Update description for governance agent template

File: governance/src/agent.js [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+
+function proposalCanceledFinding(proposalId, address, config) {
+  return Finding.fromObject({
+    name: `${config.protocolName} Governance Proposal Canceled`,
+    description: `Governance proposal ${proposalId} has been canceled`,
+    alertId: `${config.developerAbbreviation}-${config.protocolAbbreviation}-GOVERNANCE-PROPOSAL-CANCELED`,
+    type: 'Info',
+    severity: 'Info',
+    protocol: config.protocolName,
+    metadata: {
```
Comment (arbitrary-sparrow) [inline governance/src/agent.js:67]: Should `proposalId` also be included in the metadata? Also should there be an indication on whether a proposal was canceled or executed in the metadata as well?

File: governance/src/agent.js [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+function proposalExecutedFinding(proposalId, address, config) {
+  return Finding.fromObject({
+    name: `${config.protocolName} Governance Proposal Executed`,
+    description: `Governance proposal ${proposalId} has been executed`,
+    alertId: `${config.developerAbbreviation}-${config.protocolAbbreviation}-GOVERNANCE-PROPOSAL-EXECUTED`,
+    type: 'Info',
+    severity: 'Info',
+    protocol: config.protocolName,
+    metadata: {
+      address,
```
Comment (arbitrary-sparrow) [inline governance/src/agent.js:82]: Same comment as the one in `proposalCanceledFinding`

File: governance/src/utils.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  return abi;
+}
+
+function createProposalFromLog(log) {
+  const proposalId = log.args.proposalId.toString();
+  const proposal = {
+    proposalId,
+    proposer: log.args.proposer,
+    targets: log.args.targets.join(','),
+    _values: (log.args[3].map((v) => v.toString())).join(','),
```
Comment (arbitrary-sparrow) [inline governance/src/utils.js:13]: Add a comment explaining why this argument has to be parsed from the log differently

File: governance/SETUP.md (hunk 2 above)
Comment (echoAE) [reply governance/SETUP.md:4]: Done and done.

File: governance/src/agent.js (hunk 4 above)
Comment (echoAE) [reply governance/src/agent.js:67]: Done.

File: governance/src/agent.js (hunk 5 above)
Comment (echoAE) [reply governance/src/agent.js:82]: Done.

File: governance/src/agent.js [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+
+function proposalQueuedFinding(proposalId, address, config, eta) {
+  return Finding.fromObject({
+    name: `${config.protocolName} Governance Proposal Queued`,
+    description: `Governance Proposal ${proposalId} has been queued`,
+    alertId: `${config.developerAbbreviation}-${config.protocolAbbreviation}-GOVERNANCE-PROPOSAL-QUEUED`,
+    type: 'Info',
+    severity: 'Info',
+    protocol: config.protocolName,
+    metadata: {
```
[the remark this answers is not in the document]
Comment (echoAE) [reply governance/src/agent.js:95]: Done.

File: governance/package.json [hunk 8]
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "forta-{developerAbbreviation}-{protocolAbbreviation}-function-calls",
```
[the remark this answers is not in the document]
Comment (echoAE) [reply governance/package.json]: Done.

File: governance/package.json (hunk 3 above)
Comment (echoAE) [reply governance/package.json]: Done.

File: governance/SETUP.md (hunk 1 above)
Comment (echoAE) [reply governance/SETUP.md]: Done.

**Review by arbitrary-sparrow — APPROVED** [review]

**Review by arbitrary-CodeBeholder — APPROVED** [review]

**Pull request merged**